src/openforms/submissions/tokens.py

```diff
@@ -104,4 +104,95 @@ def _num_days(self, dt) -> int:
+
+    key_salt = "openforms.submissions.tokens.SubmissionTokenGenerator"
+    secret = settings.SECRET_KEY
+
+    def make_token(self, submission: Submission) -> str:
+        """
+        Return a token that can be used to verify ownership of a submission.
+        """
+        return self._make_token_with_timestamp(submission, self._num_days(date.today()))
+
+    def get_expiry_days(self, submission):
+        return settings.SUBMISSION_TOKEN_TIMEOUT_DAYS
+
+    def check_token(self, submission: Submission, token: str) -> bool:
+        """
+        Check that the token is correct for a given submission.
+        """
+
+        if not (submission and token):
+            return False
+
+        # parse the token
+        try:
+            ts_b36, _ = token.split("-")
+        except ValueError:
+            return False
+
+        try:
+            ts = base36_to_int(ts_b36)
+        except ValueError:
+            return False
+
+        # Check that the timestamp/uid has not been tampered with
+        valid_token = self._make_token_with_timestamp(submission, ts)
+        if not constant_time_compare(valid_token, token):
+            return False
+
+        expiry_days = self.get_expiry_days(submission)
+
+        # Check the timestamp is within limit. Timestamps are rounded to
+        # midnight (server time) providing a resolution of only 1 day. If a
+        # link is generated 5 minutes before midnight and used 6 minutes later,
+        # that counts as 1 day. Therefore, SUBMISSION_TOKEN_TIMEOUT_DAYS = 1 means
+        # "at least 1 day, could be up to 2."
+        if (self._num_days(date.today()) - ts) > expiry_days:
+            return False
+
+        return True
+
+    def _make_token_with_timestamp(self, submission: Submission, timestamp: int) -> str:
+        # timestamp is number of days since 2001-1-1.  Converted to
+        # base 36, this gives us a 3 digit string until about 2121
+        ts_b36 = int_to_base36(timestamp)
+        hash_string = salted_hmac(
+            self.key_salt,
+            self._make_hash_value(submission, timestamp),
+            secret=self.secret,
+        ).hexdigest()[
+            ::2
+        ]  # Limit to 20 characters to shorten the URL.
+        return "%s-%s" % (ts_b36, hash_string)
+
+    def _make_hash_value(self, submission: Submission, timestamp: int) -> str:
+        """
+        Hash the submission report ID and some state properties.
+        After a submission report has been downloaded, the state of the report is changed, so the token
+        will no longer validate.
+        Failing that, eventually settings.SUBMISSION_REPORT_URL_TOKEN_TIMEOUT_DAYS will
+        invalidate the token.
+        """
+        submission_report_attributes = ("id", "uuid")
+        submission_report_bits = [
+            str(getattr(submission, attribute) or "")
+            for attribute in submission_report_attributes
+        ]
+        return "".join(submission_report_bits) + str(timestamp)
```

Comment:
* submission report is not relevant for appointments, at all.
* the attributes also do not _change_ after the appointment has changed, and that's the whole premise of a token system like this.

After cancellation, the appointmentinfo is either deleted or updated -> _that_ should be included in the token attributes & other items that may change (appointment date/time comes to mind immediately).

This means that once an appointment is cancelled or modified, the links are now invalidated. See again the Django built in password reset token logic - tokens are invalidated after the password hash has changed there, which makes the link single-use.